Source/Libraries/GSF.Security/OIDCSecurityProvider.cs

```diff
@@ -0,0 +1,619 @@
+                    {
+                        connection.ExecuteNonQuery(database.ParameterizedQueryString("INSERT INTO AccessLog (UserName, AccessGranted) VALUES ({0}, {1})", "userName", "accessGranted"), UserData.Username, loginSuccess ? 1 : 0);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Logs information about an encountered exception to the backend data store.
+        /// </summary>
+        /// <param name="source">Source of the exception.</param>
+        /// <param name="message">Detailed description of the exception.</param>
+        /// <returns>true if logging was successful, otherwise false.</returns>
+        protected virtual bool LogError(string source, string message)
+        {
+            if (!string.IsNullOrWhiteSpace(SettingsCategory) && !string.IsNullOrWhiteSpace(source) && !string.IsNullOrWhiteSpace(message))
+            {
+                Log.Publish(MessageLevel.Error, MessageFlags.SecurityMessage, source, message);
+
+                if (UseDatabaseLogging)
+                {
+                    try
+                    {
+                        AdoDataConnection database = new AdoDataConnection(SettingsCategory);
+
+                        using (IDbConnection connection = database.Connection)
+                        {
+                            connection.ExecuteNonQuery(database.ParameterizedQueryString("INSERT INTO ErrorLog (Source, Message) VALUES ({0}, {1})", "source", "message"), source, message);
+                        }
+
+                        return true;
+                    }
+                    catch (Exception ex)
+                    {
+                        // Writing data will fail for read-only databases;
+                        // all we can do is track last exception in this case
+                        LastException = ex;
+                        Log.Publish(MessageLevel.Warning, MessageFlags.SecurityMessage, "LogErrorToDatabase", "Failed to log error to database.", "Database or ErrorLog table may be read-only or inaccessible.", ex);
+                    }
+                }
+            }
+
+            return false;
+        }
+
+       
+        /// <summary>
+        /// Exchange Authorization Code for a Token
+        /// </summary>
+        /// <param name="code"> The Authorization Code returned by the Auth Server</param>
+        private TokenResponse GetToken(string code)
+        {
+            
+            Dictionary<string, string> postParams = new Dictionary<string, string>() {
+                { "grant_type", "authorization_code" },
+                { "client_id", m_ClientID },
+                { "client_secret", m_ClientSecret },
+                { "code", code },
+                { "redirect_uri", m_RedirectURI }
+            };
+
+            void ConfigureRequest(HttpRequestMessage request)
+            {
+                request.RequestUri = new Uri($"{m_TokenEndpoint}");
```

Comment:
`m_TokenEndpoint` is already a string so the interpolation is doing nothing.

```suggestion
                request.RequestUri = new Uri(m_TokenEndpoint);
```